Add support for try with syntax

kiranandcode opened the pull request:
Add support for:
```
  try%effect comp on_complete with
  | A (s, k) -> handle_a s k
  | B (s, k) -> handle_b s k
  | C (s, k) -> handle_b s k
```

kiranandcode committed:
added support for try with syntax

cdaringe commented:
This is …amazing, and a big improvement. I’m going to merge after I pull and test in the morning. Thanks a ton for these patches

kiranandcode commented:
Full disclosure, I don't have the OCaml.4.12.0+domains switch installed locally, so I don't know if the generated code fully typechecks - the ppx may need some small tweaks to make the generated code typecheck.

I wrote the ppx such that the generated code roughly matches the previously generated code, but removed some aspects that seemed unnecessary (but could be necessary for typechecking) - in particular, I wasn't sure why your handlers all take the continue function as a parameter but then the generated code always supplies 'Obj.Effect_handlers.Deep.continue`, so I removed this parameter and just updated the test code to call `Obj.Effect_handlers.Deep.continue` directly.

Once you've had the chance to run the tests, do tell me if there are any issues.

cdaringe commented:
>  in particular, I wasn't sure why your handlers all take the continue function as a parameter but then the generated code always supplies 'Obj.Effect_handlers.Deep.continue

My thought was that if effects are to be a first class lang feature, I intentionally wanted the experience to feel like using a first class language feature, which includes not needing to import a module. I wanted to hide the modules from the user, much like the experience would be iff they were using the real syntax. I also thought it was a bit odd that effects fns came from the Obj module anyway. :)  Not a big deal at all
